components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,544 @@
+    // from your upgrade functions to signal that the database is beyond repair and can't be
+    // migrated.
+    #[error("MigrationError: {0}")]
+    MigrationError(String),
+    // Error with the migration logic struct, for example the number upgrade functions doesn't
+    // line up with start_version and end_version
+    #[error("MigrationLogicError: {0}")]
+    MigrationLogicError(String),
+    #[error("Database version too old: {0}")]
+    VersionTooOld(u32),
+    #[error("Database version too new: {0}")]
+    VersionTooNew(u32),
+    #[error("Error executing SQL: {0}")]
+    SqlError(#[from] rusqlite::Error),
+    #[error("IOError")]
+    IOError(#[from] std::io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+pub type DatabaseFunc = fn(&Connection) -> Result<()>;
+
+pub enum DatabaseLocation {
+    File(PathBuf),
+    Memory,
+}
+
+impl DatabaseLocation {
+    fn exists(&self) -> bool {
+        match self {
+            DatabaseLocation::Memory => false,
+            DatabaseLocation::File(path) => path.exists(),
+        }
+    }
+
+    fn open(&self, open_flags: OpenFlags) -> Result<Connection> {
+        match self {
+            DatabaseLocation::Memory => Ok(Connection::open_in_memory_with_flags(open_flags)?),
+            DatabaseLocation::File(path) => Ok(Connection::open_with_flags(path, open_flags)?),
+        }
+    }
+
+    fn delete(&self) -> Result<()> {
+        match self {
+            DatabaseLocation::Memory => (),
+            DatabaseLocation::File(path) => {
+                if path.exists() {
+                    std::fs::remove_file(path)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub enum ErrorHandling {
+    // Try to start over with a fresh database.  This is the behavior that most components should
+    // use (See SYNC-937, SYNC-816)
+    DeleteAndRecreate,
+    // Return an error from open function.  This allows us to potentially recover the data with an
+    // upgraded version
+    ReturnError,
+}
+
+#[derive(Clone)]
+pub struct MigrationLogic {
+    // Name to display in the logs
+    pub name: String,
+    // The first version that this migration applies to (usually 1)
+    pub start_version: u32,
+    // The version that the last upgrade function upgrades to.
+    // Note: this is intentionally redundant so that it can act as a sanity check on the length of
+    // the upgrades vec.
+    pub end_version: u32,
+    // Runs before the init/upgrade functions
+    pub prepare: Option<DatabaseFunc>,
+    // Initialize a newly created database to <end_version>
+    pub init: DatabaseFunc,
+    // Upgrade functions.  upgrades[n] will migrate version n to n+1
+    pub upgrades: Vec<DatabaseFunc>,
+    // Runs after the init/upgrade functions
+    pub finish: Option<DatabaseFunc>,
+    // How to handle migration errors
+    pub error_handling: ErrorHandling,
+}
```

Comment:
Did you consider using a `trait` for `MigrationLogic` rather than a `struct`? I don't have my head around all the details here yet, but it *might* be more convenient for consumers than using the `DatabaseFunc`-pointer approach. As a sketch, you could have a trait that provides slots for customization by the consumer, like this:

```
pub trait MigrationLogic: Clone {
    //
    // These are the parts the consumer has to fill in.
    //

    // Name to display in the logs
    const NAME: &'static str;
    // The first version that this migration applies to (usually 1)
    const START_VERSION: u32 = 1;
    // The version that the last upgrade function upgrades to.
    // Note: this is intentionally redundant so that it can act as a sanity check on the length of
    // the upgrades vec.
    const END_VERSION: u32;
    // How to handle migration errors
    const ERROR_HANDLING: ErrorHandling;

    // Initialize a newly created database to <end_version>
    fn init(&self, conn: &Connection) -> Result<()>;
    // Upgrade dispatcher. Calling upgrade_from(n) will migrate version n to n+1
    fn upgrade_from(&self, conn: &Connection, n: usize) -> Result<()>;

    //
    // These are optional extra hooks the consumer can override.
    //

    // Runs before the init/upgrade functions
    fn prepare(&self, conn: &Connection) -> Result<()> {
        Ok(())
    }
    // Runs after the init/upgrade functions
    fn finish(&self, conn: &Connection) -> Result<()> {
        Ok(())
    }    

    //
    // And these are default implementations for any `MigrationLogic`.
    //

    fn sanity_check(&self) -> Result<()> {
        // ... logic here ...
        Ok(())
    }

    fn run(&self, conn: &Connection, init: bool) -> Result<()> {
        // ... logic here ...
        Ok(())
    }

    // ... etc ...
}
```

And the consumer implementation could look like:

```
#[derive(Clone)]
pub struct AutofillMigrationLogic { }

impl MigrationLogic for AutofillMigrationLogic {
    const NAME: &'static str = "autofill db";
    const START_VERSION: u32 = 0;
    const END_VERSION: u32 = 2;
    const ERROR_HANDLING: ErrorHandling = ErrorHandling::DeleteAndRecreate;
    
    
    fn prepare(&self, conn: &Connection) -> Result<()> {
        // ... custom logic here ...
        Ok(())
    }

    
    fn init(&self, conn: &Connection) -> Result<()> {
        // ... custom logic here ...
        Ok(())
    }
    
    fn upgrade_from(&self, conn: &Connection, n: usize) -> Result<()> {
        match n {
            0 => upgrade_to_v1(conn),
            1 => upgrade_to_v2(conn),
            _ => panic!("unexpected migration version {}", n),
        }
    }

    fn finish(&self, conn: &Connection) -> Result<()> {
        // ... custom logic here ...
        Ok(())
    }  
}
```

This might help you avoid the need for declaring a specific type of `DatabaseFunc`, and make the process feel a little bit more declarative for consumers.

There are of course likely to be tradeoffs with that approach as well, and I don't feel strongly about it. Just throwing it out there as a suggestion for overall code organization.

Reply:
The one thing I wanted to avoid was having a single upgrade function than runs N upgrades.  It seemed too easy to mess up the if branches -- which is what I did with the autofill upgrade when going from v0 to v2, but it ended up working out because would always run `CREATE_SHARED_SCHEMA_SQL` at the end and we decided it was okay to lose data from the v0 databases.

But having a function that runs 1 upgrade, combined with the match seems great.  I'll refactor to use this approach.